tree_unique_args/src/loop_invariant.rs

```diff
@@ -0,0 +1,199 @@
+use std::iter;
+use strum::IntoEnumIterator;
+
+use crate::ir::{Constructor, Purpose, Sort};
+
+fn find_invariant_rule_for_ctor(ctor: Constructor) -> Option<String> {
+    let br = "\n      ";
+    let ruleset = " :ruleset always-run";
+
+    match ctor {
+        Constructor::Cons | Constructor::Nil | Constructor::Arg | Constructor::UnitExpr => None,
+        Constructor::Call => Some(format!(
+            "(rule ((find-inv-Expr loop expr) \
+            {br} (= expr (Call f arg))) \
+            {br}((find-inv-Expr loop arg)){ruleset})"
+        )),
+        Constructor::Get => Some(format!(
+            "(rule ((find-inv-Expr loop expr) \
+            {br} (= expr (Get tup i))) \
+            {br}((find-inv-Expr loop tup)){ruleset})\n \
+            (rule ((find-inv-Expr loop expr) \
+            {br} (= expr (Get (Arg id) i)) \
+            {br} (arg-inv loop i)) \
+            {br}((set (is-inv-Expr loop expr) true)){ruleset})"
+        )),
+        _ => {
+            let ctor_pattern = ctor.construct(|field| field.var());
+
+            let find_inv_ctor = ctor
+                .filter_map_fields(|field| match field.purpose {
+                    Purpose::Static(Sort::I64) | Purpose::Static(Sort::Bool) => {
+                        Some("(set (is-inv-Expr loop expr) true)".to_string())
+                    }
+                    Purpose::Static(_)
+                    | Purpose::CapturingId
+                    | Purpose::CapturedExpr
+                    | Purpose::ReferencingId => None,
+                    Purpose::SubExpr | Purpose::SubListExpr => {
+                        let var = field.var();
+                        let sort = field.sort().name();
+                        Some(format!("(find-inv-{sort} loop {var})"))
+                    }
+                })
+                .join(" ");
+
+            Some(format!(
+                "(rule ((find-inv-Expr loop expr) \
+                {br} (= expr {ctor_pattern})) \
+                {br}({find_inv_ctor}){ruleset})"
+            ))
+        }
+    }
+}
+
+fn is_invariant_rule_for_ctor(ctor: Constructor) -> Option<String> {
+    let br = "\n      ";
+    let ruleset = " :ruleset always-run";
+    let ctor_pattern = ctor.construct(|field| field.var());
+
+    match ctor {
+        // list handled in loop_invariant.egg
+        // print, read, write are not invariant
+        // assume Arg as whole is not invariant
+        Constructor::Cons
+        | Constructor::Nil
+        | Constructor::UnitExpr
+        | Constructor::Print
+        | Constructor::Read
+        | Constructor::Write
+        | Constructor::Arg => None,
+        Constructor::Get => Some(format!(
```

Comment:
Why is get special? It seems like the `_` case would work?

Reply:
Because the Get have a (Purpose::Static(Sort::I64), "i") in fields. If it pass to `_` case it generate a `(set (is-inv-Expr loop expr) true)` which is unwanted. I feel like this `i` is different from the `Purpose::Static` in `Num` and `Bool` but easy solution now is to have special case here.

Reply:
Why is it different from `Num` and `Bool`?
I think I still don't quite understand.

Reply:
Oh you are right, the match on Get in line71 is not necessary, though the match on Get on line16 is needed.